stdlib/dynarray.mli

```diff
@@ -0,0 +1,418 @@
+    The last element of [a], if not empty, is [get a (length a - 1)].
+    This operation is constant time. *)
+
+val is_empty : 'a t -> bool
+(** [is_empty a] is [true] if [a] is empty, that is, if [length a = 0]. *)
+
+val copy : 'a t -> 'a t
+(** [copy a] is a shallow copy of [a], that can be modified independently. *)
+
+
+(** {1:adding Adding elements}
+
+    Note: all operations adding elements can raise [Failure] if the
+    length would need to grow beyond {!Sys.max_array_length}. *)
+
+val add_last : 'a t -> 'a -> unit
+(** [add_last a x] adds the element [x] at the end of the array [a].
+    The length of [a] increases by [1]. *)
+
+val append_array : 'a t -> 'a array -> unit
+(** [append_array a b] adds all elements of [b] at the end of [a],
+    in the order they appear in [b].
+
+    For example, [a] will contain [1,2,3,4,5,6] after this code runs:
+    {[
+      let a = of_list [1;2;3];;
+      let () = append a [|4; 5; 6|];;
+    ]}
+*)
+
+val append_list : 'a t -> 'a list -> unit
+(** Like {!append_array} but with a list. *)
+
+val append : 'a t -> 'a t -> unit
+(** [append a b] is like [append_array a b],
+    but [b] is itself a dynamic arreay instead of a fixed-size array.
+
+    Beware! Calling [append a a] iterates on [a] and adds elements to
+    it at the same time; it is a programming error and fails with
+    [Invalid_argument].
+*)
+
+val append_seq : 'a t -> 'a Seq.t -> unit
+(** Like {!append_array} but with a sequence.
+
+    Beware! Calling [append_seq a (to_seq a)] is unspecified and may
+    result in an infinite loop, see the {!append} comment above.
+*)
+
+val append_iter :
+  'a t ->
+  (('a -> unit) -> 'iter-> unit) ->
+  'iter -> unit
+(** [append_iter a iter x] adds to [a] each element in [x]. It uses [iter]
+    to iterate over [x].
+
+    For example, [append_iter a List.iter [1;2;3]] would add elements
+    [1], [2], and [3] at the end of [a].
+    [append_iter a Queue.iter q] adds elements from the queue [q]. *)
+
+
+(** {1:removing Removing elements} *)
+
+val pop_last_opt : 'a t -> 'a option
+(** [pop_last_opt a] removes and returns the last element of [a],
+    or [None] if the array is empty. *)
+
+val pop_last : 'a t -> 'a
+(** [pop_last a] removes and returns the last element of [a].
+
+    @raise Not_found on an empty array. *)
+
+val remove_last : 'a t -> unit
+(** [remove_last a] removes the last element of [a] , or does nothing
+    if [is_empty a].
+*)
+
+val clear : 'a t -> unit
+(** [clear a] removes all the elements of [a].
+
+    It is equivalent to [truncate a 0].
+
+    Similar to {!Buffer.clear}.
+*)
+
+val truncate : 'a t -> int -> unit
+(** [truncate a n] truncates [a] to have at most [n] elements.
+
+    It removes elements whose index is great or equal than [n].
+    It does nothing if [n >= length a].
+
+    It is equivalent to:
+    {[
+      while length a > n do
+        remove_last a
+      done
+    ]}
+
+    Similar to {!Buffer.truncate}.
```

Comment:
`Buffer` is the already existing dynamic array in the stdlib, albeit specialized for bytes. Why not?

Reply:
Because these `Buffer` mentions are just chatty distractions (if not wrong as mentioned below). Doc strings should be to the point, not chatty distractions.

Why don't we add to the doc string of `List.map`, `Similar to {!Array.map}` ? And then maybe in `Array.map` we could add `Similar to {!String.map}` and in `String.map` we could add `Similar to {!List.map}`.

Reply:
Sure, all of these could point to `List.map` as the classic example. Maybe you like really short docstrings,  but beginners might like to have a bit more details or more intuition.

Reply:
Right because a beginner likely knows perfectly well the inside outs of the `Buffer` module. 

It's not that I like short doc strings, I like doc strings that are to the point for the working programmer. 

It's a balancing act but in general focusing on beginners in reference documentation – and more broadly in interaction design – is not a good idea. What works for a beginner can be very annoying and inefficient for a normal user trying to get work done. 

In the particular case of reference documentation there's plenty of other formats (tutorials, howtos, in depth explanations, etc.) you can link to which are more suited for beginners.

Reply:
I think I broadly agree with what you say :slightly_smiling_face: . Here, I think the `Buffer` comparison is not just for beginners but also for seasoned OCaml programmers for whom this module is actually new, so it anchors to a well known API.

A bit of a tangent: for me, a docstring is a bit like a commit message: the first few lines must be to the point and as informative as possible when I work (typically, when I hover the symbol). The rest, below, can be more verbose and explanatory, as it's read only if more details are needed. That's where having examples, callbacks to other modules, etc. makes sense to me.

Reply:
> For me, a docstring is a bit like a commit message: the first few lines must be to the point and as informative as possible when I work (typically, when I hover the symbol). The rest, below, can be more verbose and explanatory, as it's read only if more details are needed. That's where having examples, callbacks to other modules, etc. makes sense to me.

I tend to agree. This corresponds to the use-case where we read the docstring of one function to know more about it.
I think that @dbuenzli may have in mind more the use-case where you browse the documentation of an entire module to learn how to use it, and there having relatively shorter docstrings is more practical.

Reply:
> I think that @dbuenzli may have in mind more the use-case where you browse the documentation of an entire module to learn how to use it, and there having relatively shorter docstrings is more practical.

Not sure where you got that from. I rather have in mind people who are consulting the docs in order to achieve something. 

Each link you make or chatty blurb you add may become a distraction away from their task. 

So there should nothing superfluous, everything you add should only be here to clarify the bits of the structure item you are describing.

Reply:
(In case that was not clear. Constantly tempting me to follow links to the `Buffer` module while I'm here to understand Dynarray structure items is a distraction me)

Reply:
In general I like the idea of having a to-the-point sentence or short paragraph first, and then more explanations that may be more explanatory and potentially more "distracting". I think that if we used this for the stdlib, people in a hurry would soon learn to recognize this structure and skip the bits they don't care about. (We can help them with textual cues like "For example,", emphasis on "Warning" if there needs any, and careful placement of the more important information first.)

Stdlib has a special status in my mind because it is a library that most people use at the same time as they learn the language, so it gets more beginners looking at its doc than most other packages. (Some of those beginners may be familiar with other languages, but not necessarily with OCaml style, conventions, etc.) I'm especially interested in proposing well-thought-out usage examples for various functions, ideally all of them.

In the specific case of this discussion, I went along Daniel's suggestion and removed the references to the Buffer module. Buffer is a useful reference when designing the API and we should strive to remain consistent with it. For the rest, people familiar with Buffer will see the connection, but people unfamiliar with Buffer will not be helped much by cross-referencing its functions here I think, the mention of Buffer in the documentation preambule was probably enough.

Reply:
> Stdlib has a special status in my mind because it is a library that most people use at the same time as they learn the language, so it gets more beginners looking at its doc than most other packages. (Some of those beginners may be familiar with other languages, but not necessarily with OCaml style, conventions, etc.) 

Just to mention, I would be exactly of the opposite view :–) 

I guess what you say makes sense from an academic point of view but lots of people, especially those who already know how to program, learn the language by doing something with it. 

This means they will rather fall on it as a side effect of reading the documentation of *other packages*. (The internet law of user experience I constantly need to remind my clients: "Users spend most of their time on other sites" :–)

If Stdlib really has a special status then it should rather be that, being the Stdlib:

1. It will perused a lot by the working programmers over the years, so it should be particularly efficient for them
2. It should be exemplar in setting the style we want in OCaml. Users are likely to copy-cat the documentation style. In fact that's where I was taught to write doc strings the way I usually suggest (well in some parts, since the style is a tad inconsistent across the stdlib).
 
> I'm especially interested in proposing well-thought-out usage examples for various functions, ideally all of them.

I think making short commented examples to cut and paste at the end of modules to show how to use a *conjunction of functions* to achieve a simple goal is much better than trying to add examples in each doc string. 

I know people are very keen in having stuff like `List.map succ [1; 2; 3] = [2; 3; 4]` in doc strings but in my experience people tend to derive incomplete specifications in their head from them which leads to bugs (because they are in a hurry, *see* it seems to do mostly what they are looking for without thinking/seeing the edge cases). 

For me doc strings should tell you how to read and understand code that uses the item -hence the `[f …] is` style and equations – not how it behaves on particular examples (except perhaps to highlight a few tricky edge cases).